include/nbl/asset/utils/CDirQuantCacheBase.h

```diff
@@ -397,29 +413,42 @@ class CDirQuantCacheBase : public virtual core::IReferenceCounted, public impl::
 				{
 					const auto fit = findBestFit<dimensions,quantizationBits>(absValue);
 
-					quantized = abs(fit);
+					const auto abs_fit = to_float32_t4(abs(fit));
+          quantized = hlsl::uint32_t4(abs_fit.x, abs_fit.y, abs_fit.z, abs_fit.w);
+
 					insertIntoCache<CacheFormat>(key,quantized);
 				}
 			}
 
-			//return quantized.
-			const auto negativeMulVec = hlsl::float32_t3(negativeMask.x ? -1 : 1, negativeMask.y ? -1 : 1, negativeMask.z ? -1 : 1);
-      return value_type_t<CacheFormat>(negativeMulVec * quantized.getValue());
+			auto switch_vec = [](hlsl::uint32_t4 val1, hlsl::uint32_t4 val2, hlsl::bool4 mask)
+      {
+					hlsl::uint32_t4 retval;
+					retval.x = mask.x ? val2.x : val1.x;
+					retval.y = mask.y ? val2.y : val1.y;
+					retval.z = mask.z ? val2.z : val1.z;
+					retval.w = mask.w ? val2.w : val1.w;
+					return retval;
+      };
+
+      const hlsl::uint32_t4 xorflag((0x1u << (quantizationBits + 1u)) - 1u);
+      auto restoredAsVec = quantized.getValue() ^ switch_vec(hlsl::uint32_t4(0u), hlsl::uint32_t4(xorflag), negativeMask);
+      restoredAsVec += switch_vec(hlsl::uint32_t4(0u), hlsl::uint32_t4(1u), negativeMask);
```

Comment:
if `negativeMask` was a `uint32_t4` you could just write `B&negativeMask` instead of `switch_vec(A,B,negativeMask` because your A is always 0